Displays hardware information and identifies incompatible builds

grahampugh opened the pull request:
Hi Greg

To start the process of breaking down my updates to this script to see if you are interested in merging it, I have created a new branch. 

In this branch, I have merged in all your upstream changes regarding catalog choice and post date, and added the part of my `master` fork which identifies the current hardware on which the script is being run. The initial output of the script is therefore as in this example:

```
$ sudo python installinstallmacos.py 

installinstallmacos.py - get macOS installers from the Apple software catalog

This Mac:
Model Identifier : MacBookPro14,2
Board ID         : Mac-CAD6701F7CEA0921
OS Version       : 10.14
Build ID         : 18A391
```

Note that the script also checks for a VM. This results in an additional line when a VM is identified: 

```
Identified as a Virtual Machine
``` 

I added a new column to the list of items called `Notes`, which displays a string if the listed build is incompatible with the current hardware, either of:

* `Unsupported Model Identifier`
* `Unsupported macOS version`
* `Unsupported Board ID`

This results in an output as in this example run on a T2 MBP running macOS 10.14.0:

```
 #  ProductID       Version    Build    Post Date   Title                          Notes
 1  041-08708       10.14      18A391   2018-09-24  macOS Mojave                   
 2  091-93471       10.14.1    18B75    2018-10-30  macOS Mojave                   
 3  091-94326       10.13.6    17G65    2018-07-11  macOS High Sierra              Unsupported macOS version
 4  091-95774       10.13.6    17G2208  2018-07-30  macOS High Sierra              Unsupported Board ID
```

In this branch's commit, the script then proceeds as normal. 

These hardware checks are the groundwork for adding the ability to automate the correct build based on some additional arguments. I have not added those into the PR yet, so as to maintain clarity, and will await feedback on these changes before adding them. I leave the other PR in place, in case you want to cross-reference what I'm ultimately trying to achieve.

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
 
+def get_board_id():
+    '''Gets the local system board ID'''
+    ioreg_cmd = ['ioreg', '-p', 'IODeviceTree', '-r', '-n', '/', '-d', '1']
```
full path to binary, please.

grahampugh replied:
Done

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
+                board_id = board_id[board_id.find('<"')+2:board_id.find('">')]
+                return board_id
+    except subprocess.CalledProcessError, err:
+        raise ReplicationError(err)
```
ReplicationError seems the wrong error to raise -- this has nothing to do with replicating the files from Apple

grahampugh replied:
Changed to a simpler output

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
+            if cpu_features[i] == "VMM":
+                is_vm = True
+    except subprocess.CalledProcessError, err:
+        raise ReplicationError(err)
```
ReplicationError seems the wrong error to raise -- this has nothing to do with replicating the files from Apple

grahampugh replied:
Changed to a simpler output

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
+        sysctl_output = subprocess.check_output(sysctl_cmd)
+        hw_model = sysctl_output.split(" ")[-1].split("\n")[0]
+    except subprocess.CalledProcessError, err:
+        raise ReplicationError(err)
```
ReplicationError seems the wrong error to raise -- this has nothing to do with replicating the files from Apple

grahampugh replied:
Changed to a simpler output

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
+            if 'BuildVersion' in line:
+                build_info.insert(1, line.split("\t")[-1])
+    except subprocess.CalledProcessError, err:
+        raise ReplicationError(err)
```
ReplicationError seems the wrong error to raise -- this has nothing to do with replicating the files from Apple

grahampugh replied:
Changed to a simpler output

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
         for key, value in seed_catalogs.items():
             if sucatalog_url == value:
                 return key
+        return ''
```
This is distracting because this code is _not_ a change from the current master branch.

grahampugh replied:
I've no idea why this is highlighted. As far as I can tell there is no difference. I see it sometimes in my own PRs but not sure how to deal with it except just reject it during the merge.

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,110 @@
+def get_default_catalog():
+    '''Returns the default softwareupdate catalog for the current OS'''
+    darwin_major = os.uname()[2].split('.')[0]
+    return DEFAULT_SUCATALOGS.get(darwin_major)
```
again this is all in master -- not clear why this highlighted as a change.

grahampugh replied:
As above.

gregneagle replied:
OK, but complicates reviewing your changes since I have to double-check against master to see what's "really" changed.

gregneagle commented on installinstallmacos.py:
```diff
@@ -269,6 +342,30 @@ def parse_dist(filename):
+            line = line.rstrip()  # remove '\n' at end of line
+            if 'boardIds' in line:
+                supported_board_ids = line.split(" ")[-1][:-1]
+                return supported_board_ids
```
This seems extremely fragile. the dist file is XML -- you should be parsing it using an XML parser.
(This also sets aside the possibility that Apple could change any of this at any time)

grahampugh replied:
I'm not sure what we would gain there. Although it's in an XML tree, it's actually in embedded JavaScript. So, sure, I could parse the output down to the specific `<scripts>` key but no further than that using XML parsing.  I've added the requirement for `var boardIds =` so it couldn't catch another instance of `boardIds`. Any other suggestions to make this more robust? Or just check that `board_id` has a value and ignore if not?

gregneagle replied:
At least make the function return something if 'var boardIds =' is never found -- ideally if Apple changed the format of the script, the code would just act as before you added any of this.

gregneagle commented on installinstallmacos.py:
```diff
@@ -269,6 +342,30 @@ def parse_dist(filename):
+    '''Parses a softwareupdate dist file, returning a list of non-supported
+    ModelIdentifiers'''
+    unsupported_models = ""
+    with open(filename) as search:
```
Same comment about XML parsing.

grahampugh replied:
Same problem as above.

gregneagle replied:
So same comment as above. If 'nonSupportedModels' is never found, the function will return None

What happens if nonSupportedModels are broken over several lines?

gregneagle commented on installinstallmacos.py:
```diff
@@ -415,16 +564,26 @@ def main():
+            validation_string = 'Unsupported Board ID'
+        elif get_lowest_version(build_info[0],product_info[product_id]['version']) != build_info[0]:
+            validation_string = 'Unsupported macOS version'
+
```
I think it should be clearer that an item is unsupported _on this Mac_

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,113 @@
+    try:
+        sysctl_output = subprocess.check_output(sysctl_cmd)
+        cpu_features = sysctl_output.split(" ")
+        is_vm = False
```
Move this before the try block so that if you do get an exception the function will return False instead of throwing another exception.

grahampugh replied:
Done

gregneagle commented on installinstallmacos.py:
```diff
@@ -30,42 +30,114 @@
+                return board_id
+    except subprocess.CalledProcessError, err:
+        print "Board ID could not be identified. Error was: %s" % err
+        return ''
```
Still can fall through and return None if 'board-id' is never found.

grahampugh replied:
Could I just change this to a non-specific `except:`?

gregneagle replied:
That would not fix the issue I'm talking about, which is simply not finding "board-id" in the output. Perhaps this would never ever happen. But it's easy to make the code behave better in case it does.

grahampugh replied:
I think I resolved this now. None of the definitions that I've added would now return `None` if they failed to identify the value they are looking for. They now return `Not identified` so that the outputted text is informative.

gregneagle commented:
I'm having a hard time now following your subsequent changes, but I'm trying...

grahampugh commented:
I've reworked the `board_id` and `unsupported_model` definitions into a single def, which now uses minidom (since it's already used, rather than adding ElementTree) to parse out the script objects. There's actually presently two, once of which is empty, so I loop through them and look for the one that contains the board ID list and the unsupported model list (currently the same script object).

I still have to parse by line to obtain these lists. As far as I know, you don't split lines in JavaScript unlike in python, so unless Apple change their format significantly, we should be OK. Of course, they could change anything in the XML, or move to plist or json, etc...

There could be some additional use in this rework, since there are now proper lists for supported_board_ids and unsupported_model_types, for instance we could output them to a plist along with the other data, for external use.

I also added a check for content in each of these lists. If they're empty, they do not generate any incompatibility message.

grahampugh commented:
Withdrawing the PR.
